dm/worker/worker.go

```diff
@@ -71,14 +71,29 @@ type Worker struct {
 }
 
 // NewWorker creates a new Worker
-func NewWorker(cfg *Config) (*Worker, error) {
-	w := &Worker{
+func NewWorker(cfg *Config) (w *Worker, err error) {
+	w = &Worker{
 		cfg:           cfg,
 		relayHolder:   NewRelayHolder(cfg),
 		tracer:        tracing.InitTracerHub(cfg.Tracer),
 		subTaskHolder: newSubTaskHolder(),
 		l:             log.With(zap.String("component", "worker controller")),
 	}
+	w.ctx, w.cancel = context.WithCancel(context.Background())
+
+	defer func(w2 *Worker) {
+		if err != nil { // when err !=nil, `w` will become nil in this func, so we pass `w` in defer.
+			// release resources, NOTE: we need to refactor New/Init/Start/Close for components later.
+			w2.cancel()
+			w2.subTaskHolder.closeAllSubTasks()
+			if w2.meta != nil {
+				w2.meta.Close()
+			}
+			if w2.db != nil {
+				w2.db.Close()
+			}
```

Comment:
can we just use w2.Close(), and maybe need add some check in `Close` (close something when it is not nil)

Reply:
I think we can.
But I also think we should refactor the pattern of components' new/init/start/close later. 
like `new` only creates fields but does not acquire (IO) resources and start no goroutines, then when `new` failed, no operation needs to do and let GC to remove the objects is enough.

I think we need to do some refactoring when developing the HA version of DM.

what's your opinion?

Reply:
OK，we can do this later